File: mpi/docs/mpi_logical/example_1.f90
!> @brief Illustrate how to communicate a logical between 2 MPI processes.
!> @details This application is meant to be run with 2 MPI processes: 1 sender
!> and 1 receiver. The former sends a logical to the latter, which prints it.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: size
    INTEGER, PARAMETER :: sender_rank = 0
    INTEGER, PARAMETER :: receiver_rank = 1
    INTEGER :: my_rank
    LOGICAL :: boolToSend
    LOGICAL :: boolReceived

    CALL MPI_Init(ierror)

    ! Check that 2 MPI processes are used.
    CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierror)
    IF (size .NE. 2) THEN
        WRITE(*,'(A)') 'This application is meant to be run with 2 MPI processes.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierror)
    END IF

    ! Get my rank and do the corresponding job.
    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)
    SELECT CASE (my_rank)
        CASE (sender_rank)
            ! Send the logical
            boolToSend = .TRUE.
            WRITE(*,'(A,I0,A,L0,A)') '[MPI process ', my_rank, '] I send logical: ', boolToSend, '.'
            CALL MPI_Ssend(boolToSend, 1, MPI_LOGICAL, receiver_rank, 0, MPI_COMM_WORLD, ierror)
        CASE (receiver_rank)
            ! Receive the logical
            CALL MPI_Recv(boolReceived, 1, MPI_LOGICAL, sender_rank, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror)
            WRITE(*,'(A,I0,A,L0,A)') '[MPI process ', my_rank, '] I received logical: ', boolReceived, '.'
    END SELECT

    CALL MPI_Finalize(ierror)
END PROGRAM main

File: mpi/docs/mpi_uint32_t/example_1.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <mpi.h>

/**
 * @brief Illustrate how to communicate an unsigned 4-byte int between 2 MPI
 * processes.
 * @details This application is meant to be run with 2 MPI processes: 1 sender
 * and 1 receiver. The former sends an unsigned 4-byte int to the latter, which
 * prints it.
 **/
int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	// Check that 2 MPI processes are used.
	int size;
	MPI_Comm_size(MPI_COMM_WORLD, &size);
	if(size != 2)
	{
		printf("This application is meant to be run with 2 MPI processes.\n");
		MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
	}

	// Get my rank and do the corresponding job.
	enum role_ranks { SENDER, RECEIVER };
	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
	switch(my_rank)
	{
		case SENDER:
		{
			// Send the int
			uint32_t intToSend = 123456789;
			printf("[MPI process %d] I send unsigned int: %u.\n", my_rank, intToSend);
			MPI_Ssend(&intToSend, 1, MPI_UINT32_T, RECEIVER, 0, MPI_COMM_WORLD);
			break;
		}
		case RECEIVER:
		{
			// Receive the int
			uint32_t intReceived;
			MPI_Recv(&intReceived, 1, MPI_UINT32_T, SENDER, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			printf("[MPI process %d] I received unsigned int: %u.\n", my_rank, intReceived);
			break;
		}
	}

	MPI_Finalize();

	return EXIT_SUCCESS;
}
